### PR TITLE
Users can now be copied form other courses. Fixes #824

### LINKED ISSUE #824: Copy participants from other course
It should be possible to not only import participants to a course from an excel sheet, but also to copy the participants from another course.

### PR DESCRIPTION
**Pull request by schumannd**
Fixes #824

### CODE REVIEW

File: evap/staff/views.py [hunk 1]
```diff
@@ -635,11 +635,16 @@ def course_participant_import(request, semester_id, course_id):
...
         test_run = operation == 'test'
 
-        # Parse table.
-        imported_users = UserImporter.process(request, excel_file, test_run)
+        # Import from Excel file
+        if excel_file:
+            imported_users = UserImporter.process(request, excel_file, test_run)
+        else:
+            imported_users = import_course.participants.all()
+            messages.success(request, "%d Participants would be added to course %s" % (len(imported_users), course.name))
```
Comment (janno42) [inline evap/staff/views.py]: this message should only be shown when testing. when actually adding the participants another success message is already shown.

File: evap/staff/forms.py [hunk 2]
```diff
@@ -31,7 +31,14 @@ class ImportForm(forms.Form):
 
 
 class UserImportForm(forms.Form):
-    excel_file = forms.FileField(label=_("Excel file"))
+    course = forms.ModelChoiceField(Course.objects.all(), required=False, label=_("Import from Course:"))
+    excel_file = forms.FileField(label=_("Import from Excel file: "), required=False)
+
+    def clean(self):
+        if self.cleaned_data['course'] and self.cleaned_data['excel_file']:
+            raise ValidationError('Please only set course OR Excel File')
```
Comment (janno42) [inline evap/staff/forms.py]: please change to `Please select only one of course or Excel file.`

File: evap/staff/forms.py [hunk 3]
```diff
@@ -31,7 +31,14 @@ class ImportForm(forms.Form):
...
+        if not self.cleaned_data['course'] and not self.cleaned_data['excel_file']:
+            raise ValidationError('Please set either course OR Excel File')
```
Comment (janno42) [inline evap/staff/forms.py]: please change to `Please select either course or Excel file.`

File: evap/staff/forms.py [hunk 4]
```diff
@@ -31,7 +31,14 @@ class ImportForm(forms.Form):
 
 
 class UserImportForm(forms.Form):
-    excel_file = forms.FileField(label=_("Excel file"))
+    course = forms.ModelChoiceField(Course.objects.all(), required=False, label=_("Import from Course:"))
```
Comment (janno42) [inline evap/staff/forms.py]: Please make this `Copy from Course`

File: evap/staff/forms.py [hunk 5]
```diff
@@ -31,7 +31,14 @@ class ImportForm(forms.Form):
...
+    excel_file = forms.FileField(label=_("Import from Excel file: "), required=False)
```
Comment (janno42) [inline evap/staff/forms.py]: Please remove the colon

**Comment by schumannd** [comment]
All TODOs done

File: evap/staff/forms.py [hunk 6]
```diff
@@ -31,7 +31,22 @@ class ImportForm(forms.Form):
 
 
 class UserImportForm(forms.Form):
-    excel_file = forms.FileField(label=_("Excel file"))
+    excel_file = forms.FileField(label=_("Import from Excel file"), required=False)
+    course = forms.ModelChoiceField(Course.objects.all(), required=False, label=_("Copy from Course"))
+
+    choices = [('', '---------')]
+    for semester in Semester.objects.all():
```
Comment (karyon) [inline evap/staff/forms.py]: this is evaluated at module import time, and apparently stuff hasn't been initialized completely at that point.

File: evap/staff/forms.py [hunk 7]
```diff
@@ -31,7 +31,24 @@ class ImportForm(forms.Form):
 
 
 class UserImportForm(forms.Form):
-    excel_file = forms.FileField(label=_("Excel file"))
+    excel_file = forms.FileField(label=_("Import from Excel file"), required=False)
+    course = forms.ModelChoiceField(Course.objects.all(), required=False, label=_("Copy from Course"))
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        choices = [('', '---------')]
```
Comment (karyon) [inline evap/staff/forms.py]: is this still necessary?

also, please add some comment explaining the grouping thing

File: evap/staff/views.py [hunk 8]
```diff
@@ -635,11 +635,17 @@ def course_participant_import(request, semester_id, course_id):
...
 
-        # Parse table.
-        imported_users = UserImporter.process(request, excel_file, test_run)
+        # Import from Excel file
+        if excel_file:
+            imported_users = UserImporter.process(request, excel_file, test_run)
+        else:
+            imported_users = import_course.participants.all()
+            if test_run:
+                messages.success(request, "%d Participants would be added to course %s" % (len(imported_users), course.name))
```
Comment (karyon) [inline evap/staff/views.py]: can't we do that for both the excel file and the course copy thing?

File: evap/staff/templates/staff_course_participant_import.html [hunk 9]
```diff
@@ -6,7 +6,7 @@
 {% block content %}
     {{ block.super }}
     <p>
-        {% trans "Upload course participants data from excel file" %} (<a href="{% get_static_prefix %}sample_user.xls">{% trans "Sample File"%}</a>). {% trans "This will create user accounts for all contained students and adds them as participants for the course." %}
+        {% trans "Copy participants from another course or import participants data from Excel file" %} (<a href="{% get_static_prefix %}sample_user.xls">{% trans "Sample File"%}</a>). {% trans "The latter will create user accounts for all contained students if necessary before adding them as participants for the course." %}
```
Comment (karyon) [inline evap/staff/templates/staff_course_participant_import.html]: isn't it rather "participant data"?

Comment (janno42) [inline evap/staff/templates/staff_course_participant_import.html]: yes

File: evap/staff/forms.py [hunk 10]
```diff
@@ -31,7 +31,27 @@ class ImportForm(forms.Form):
...
 
 class UserImportForm(forms.Form):
-    excel_file = forms.FileField(label=_("Excel file"))
+    excel_file = forms.FileField(label=_("Import from Excel file"), required=False)
+    course = forms.ModelChoiceField(Course.objects.all(), empty_label='<empty>', required=False, label=_("Copy from Course"))
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        # Here we Split the courses by Semester and create Supergroups for them. We also make sure to include an empty option.
```
Comment (karyon) [inline evap/staff/forms.py]: lowercase all the S

File: evap/staff/forms.py [hunk 11]
```diff
@@ -31,7 +31,27 @@ class ImportForm(forms.Form):
...
+        choices = [('', '<empty>')]
+        for semester in Semester.objects.all():
+            course_choices = [(course.pk, course.name) for course in Course.objects.filter(semester=semester)]
+            if course_choices:
+                choices += [(semester.name, course_choices)]
+
+        # THis is the hacky part where we set the private property of the ModelChoiceField
+        self.fields['course']._choices = choices
```
Comment (karyon) [inline evap/staff/forms.py]: the comment is not helpful right now, it's just re-stating what the code does. either say *why* you needed to do the hack or remove it.

also, i've seen multiple stackoverflow posts where people wrote into `self.fields['foo'].choices`, without the underscore, does that not work?

File: evap/staff/templates/staff_course_participant_import.html [hunk 12]
```diff
@@ -6,7 +6,7 @@
 {% block content %}
     {{ block.super }}
     <p>
-        {% trans "Upload course participants data from excel file" %} (<a href="{% get_static_prefix %}sample_user.xls">{% trans "Sample File"%}</a>). {% trans "This will create user accounts for all contained students and adds them as participants for the course." %}
+        {% trans "Copy participant data from another course or import participants data from Excel file" %} (<a href="{% get_static_prefix %}sample_user.xls">{% trans "Sample File"%}</a>). {% trans "The latter will create user accounts for all contained students if necessary before adding them as participants for the course." %}
```
Comment (karyon) [inline evap/staff/templates/staff_course_participant_import.html]: `participant data` plz

File: evap/staff/forms.py (hunk 11 above)
Comment (schumannd) [inline evap/staff/forms.py]: you are right

**Pull request merged**